rattan-core/src/cells/token_bucket.rs

```diff
@@ -0,0 +1,1160 @@
+                self.peak_token_bucket.burst = minburst;
+            }
+            (Some(peakrate), Some(minburst)) => {
+                self.peak_token_bucket.rate = peakrate;
+                self.peak_token_bucket.burst = minburst;
+            }
+        };
+        self.peak_token_bucket.tokens = length_to_time(
+            self.peak_token_bucket.burst.as_u64(),
+            self.peak_token_bucket.rate,
+        );
+    }
+
+    fn set_config(&mut self, config: TokenBucketCellConfig) {
+        let now = Instant::now();
+        debug!(
+            "set_config: Previous next_available distance: {:?}",
+            self.next_available - now
+        );
+        debug!("Previous check_point distance: {:?}", now - self.last_check);
+
+        self.last_check = now;
+
+        // modify self.token_bucket
+        self.set_token_bucket(config.rate, config.burst);
+        self.set_peak_token_bucket(config.peakrate, config.minburst);
+
+        debug!(
+            "toks: {}, ptoks: {}",
+            self.token_bucket.tokens.as_nanos(),
+            self.peak_token_bucket.tokens.as_nanos()
+        );
+
+        // calculate max_size
+        self.max_size = min(self.token_bucket.burst, self.peak_token_bucket.burst);
+
+        // set queue
+        if let Some(limit) = config.limit {
+            debug!(?limit, "Set inner queue byte_limit: ");
+            self.packet_queue.configure(DropTailQueueConfig::new(
+                None,
+                limit,
+                crate::cells::bandwidth::BwType::NetworkLayer,
+            ));
+        }
+
+        // drop all packets larger than max_size in the queue
+        self.packet_queue
+            .drop_large_packet(self.max_size.0 as usize);
+
+        // set next_available
+        if !self.packet_queue.is_empty() {
+            debug!("set_config: set next_available to now");
+            self.next_available = now;
+        } else {
+            self.next_available = now + LARGE_DURATION;
+        }
+    }
+
+    // calculate and return next_available Instant
+    fn calculate_next_available(&self, now: Instant, size: ByteSize) -> Instant {
+        let consumed_tokens = length_to_time(size.as_u64(), self.token_bucket.rate);
+        let consumed_ptokens = length_to_time(size.as_u64(), self.peak_token_bucket.rate);
+        if self.token_bucket.tokens >= consumed_tokens
+            && self.peak_token_bucket.tokens >= consumed_ptokens
+        {
+            debug!("calculate: set next_available to now");
+            now
+        } else if self.token_bucket.tokens >= consumed_tokens {
+            now + (consumed_ptokens - self.peak_token_bucket.tokens)
+        } else if self.peak_token_bucket.tokens >= consumed_ptokens {
+            now + (consumed_tokens - self.token_bucket.tokens)
+        } else {
+            now + max(
+                consumed_tokens - self.token_bucket.tokens,
+                consumed_ptokens - self.peak_token_bucket.tokens,
+            )
+        }
```

Comment:
You can avoid too much else-if statements by using [saturating_sub](https://doc.rust-lang.org/std/time/struct.Duration.html#method.saturating_sub). The following is an example of replacing the whole if-else block, you can check if it is correct to use:

```rust
now + max(consumed_tokens.saturating_sub(self.token_bucket.tokens), consumed_ptokens.saturating_sub(self.peak_token_bucket.tokens))
```

If the calculation can be exactly simplified to such one sentence, then you can just remove this function, since `comsumed_xxx` is always available where you call this function.